### ISSUE TITLE
barlow tracking fails during wbfm pipline Error(s) in loading state_dict for BarlowTwins3d: 	size mismatch for projector.6.weight: copying a param with shape torch.Size([256, 2048]) from checkpoint, the shape in current model is torch.Size([2048, 2048])

### ISSUE THREAD

**Issue by Itamar-Lev** [Zimmer-lab/barlow_track#67]
RuntimeError in file /lisc/data/scratch/neurobiology/zimmer/ItamarLev/WBFM/WBFM_projects/63x_test/2025-07-23_15-24_control_1per_63x_oil_test_worm1-2025-07-23/snakemake/pipeline.smk, line 68:
Error(s) in loading state_dict for BarlowTwins3d:
	size mismatch for projector.6.weight: copying a param with shape torch.Size([256, 2048]) from checkpoint, the shape in current model is torch.Size([2048, 2048]).
  File "/lisc/data/scratch/neurobiology/zimmer/ItamarLev/WBFM/WBFM_projects/63x_test/2025-07-23_15-24_control_1per_63x_oil_test_worm1-2025-07-23/snakemake/pipeline.smk", line 221, in __rule_barlow_tracking
  File "/lisc/data/scratch/neurobiology/zimmer/ItamarLev/WBFM/WBFM_projects/63x_test/2025-07-23_15-24_control_1per_63x_oil_test_worm1-2025-07-23/snakemake/pipeline.smk", line 68, in _run_helper
  File "/lisc/scratch/neurobiology/zimmer/.conda/envs/wbfm/lib/python3.8/site-packages/sacred/experiment.py", line 277, in run
  File "/lisc/scratch/neurobiology/zimmer/.conda/envs/wbfm/lib/python3.8/site-packages/sacred/run.py", line 238, in __call__
  File "/lisc/scratch/neurobiology/zimmer/.conda/envs/wbfm/lib/python3.8/site-packages/sacred/config/captured_function.py", line 42, in captured_function
  File "/lisc/scratch/neurobiology/zimmer/wbfm/code/wbfm/wbfm/scripts/pipeline_alternate/3-track_using_barlow.py", line 43, in main
  File "/lisc/scratch/neurobiology/zimmer/wbfm/code/barlow_track/barlow_track/utils/track_using_barlow.py", line 117, in track_using_barlow_from_config
  File "/lisc/scratch/neurobiology/zimmer/wbfm/code/barlow_track/barlow_track/utils/barlow.py", line 368, in load_barlow_model
  File "/lisc/scratch/neurobiology/zimmer/.conda/envs/wbfm/lib/python3.8/site-packages/torch/nn/modules/module.py", line 2152, in load_state_dict
  File "/lisc/scratch/neurobiology/zimmer/.conda/envs/wbfm/lib/python3.8/concurrent/futures/thread.py", line 57, in run
ERROR:snakemake.logging:RuleException:
RuntimeError in file /lisc/data/scratch/neurobiology/zimmer/ItamarLev/WBFM/WBFM_projects/63x_test/2025-07-23_15-24_control_1per_63x_oil_test_worm1-2025-07-23/snakemake/pipeline.smk, line 68:
Error(s) in loading state_dict for BarlowTwins3d:
	size mismatch for projector.6.weight: copying a param with shape torch.Size([256, 2048]) from checkpoint, the shape in current model is torch.Size([2048, 2048]).
  File "/lisc/data/scratch/neurobiology/zimmer/ItamarLev/WBFM/WBFM_projects/63x_test/2025-07-23_15-24_control_1per_63x_oil_test_worm1-2025-07-23/snakemake/pipeline.smk", line 221, in __rule_barlow_tracking
  File "/lisc/data/scratch/neurobiology/zimmer/ItamarLev/WBFM/WBFM_projects/63x_test/2025-07-23_15-24_control_1per_63x_oil_test_worm1-2025-07-23/snakemake/pipeline.smk", line 68, in _run_helper
  File "/lisc/scratch/neurobiology/zimmer/.conda/envs/wbfm/lib/python3.8/site-packages/sacred/experiment.py", line 277, in run
  File "/lisc/scratch/neurobiology/zimmer/.conda/envs/wbfm/lib/python3.8/site-packages/sacred/run.py", line 238, in __call__
  File "/lisc/scratch/neurobiology/zimmer/.conda/envs/wbfm/lib/python3.8/site-packages/sacred/config/captured_function.py", line 42, in captured_function
  File "/lisc/scratch/neurobiology/zimmer/wbfm/code/wbfm/wbfm/scripts/pipeline_alternate/3-track_using_barlow.py", line 43, in main
  File "/lisc/scratch/neurobiology/zimmer/wbfm/code/barlow_track/barlow_track/utils/track_using_barlow.py", line 117, in track_using_barlow_from_config
  File "/lisc/scratch/neurobiology/zimmer/wbfm/code/barlow_track/barlow_track/utils/barlow.py", line 368, in load_barlow_model
  File "/lisc/scratch/neurobiology/zimmer/.conda/envs/wbfm/lib/python3.8/site-packages/torch/nn/modules/module.py", line 2152, in load_state_dict
  File "/lisc/scratch/neurobiology/zimmer/.conda/envs/wbfm/lib/python3.8/concurrent/futures/thread.py", line 57, in run

**Comment by Charles-Fieseler-Vienna** [comment]
This is because I've been changing the default values of certain layer sizes... If you want to start from a pretrained network all the later sizes must be the same

An easy solution is to train from scratch, meaning delete the path to the pretrained model

**Comment by xmryyx** [comment]
I trained a new barlow tracker and wanted to test it on a subset of data that hasn't been analyzed before.
Even though I analyzed them from scratch I get the same error.

ERROR:snakemake.logging:Error in rule barlow_embedding:
    jobid: 0
    input: /lisc/data/scratch/neurobiology/zimmer/MariaBoehm/data/WBFM/projects/55_beads_barlow/2025-10-03_17-21_beads_55per_worm12-2025-10-03/project_config.yaml, /lisc/data/scratch/neurobiology/zimmer/MariaBoehm/data/WBFM/projects/55_beads_barlow/2025-10-03_17-21_beads_55per_worm12-2025-10-03/1-segmentation/metadata.pickle
    output: /lisc/data/scratch/neurobiology/zimmer/MariaBoehm/data/WBFM/projects/55_beads_barlow/2025-10-03_17-21_beads_55per_worm12-2025-10-03/3-tracking/barlow_tracker/worm_tracker_barlow.pickle

RuleException:
RuntimeError in file /lisc/data/scratch/neurobiology/zimmer/MariaBoehm/data/WBFM/projects/55_beads_barlow/2025-10-03_17-21_beads_55per_worm12-2025-10-03/snakemake/pipeline.smk, line 69:
Error(s) in loading state_dict for BarlowTwins3d:
	size mismatch for projector.6.weight: copying a param with shape torch.Size([256, 2048]) from checkpoint, the shape in current model is torch.Size([2048, 2048]).
  File "/lisc/data/scratch/neurobiology/zimmer/MariaBoehm/data/WBFM/projects/55_beads_barlow/2025-10-03_17-21_beads_55per_worm12-2025-10-03/snakemake/pipeline.smk", line 253, in __rule_barlow_embedding
  File "/lisc/data/scratch/neurobiology/zimmer/MariaBoehm/data/WBFM/projects/55_beads_barlow/2025-10-03_17-21_beads_55per_worm12-2025-10-03/snakemake/pipeline.smk", line 69, in _run_helper
  File "/lisc/scratch/neurobiology/zimmer/.conda/envs/wbfm/lib/python3.8/site-packages/sacred/experiment.py", line 277, in run
  File "/lisc/scratch/neurobiology/zimmer/.conda/envs/wbfm/lib/python3.8/site-packages/sacred/run.py", line 238, in __call__
  File "/lisc/scratch/neurobiology/zimmer/.conda/envs/wbfm/lib/python3.8/site-packages/sacred/config/captured_function.py", line 42, in captured_function
  File "/lisc/scratch/neurobiology/zimmer/wbfm/code/wbfm/wbfm/scripts/pipeline_alternate/3-embed_using_barlow.py", line 43, in main
  File "/lisc/scratch/neurobiology/zimmer/wbfm/code/barlow_track/barlow_track/utils/track_using_barlow.py", line 132, in embed_using_barlow_from_config
  File "/lisc/scratch/neurobiology/zimmer/wbfm/code/barlow_track/barlow_track/utils/barlow.py", line 385, in load_barlow_model
  File "/lisc/scratch/neurobiology/zimmer/.conda/envs/wbfm/lib/python3.8/site-packages/torch/nn/modules/module.py", line 2152, in load_state_dict
  File "/lisc/scratch/neurobiology/zimmer/.conda/envs/wbfm/lib/python3.8/concurrent/futures/thread.py", line 57, in run


Here is my snakemake_config.yaml:

# This file contains the default values for the snakemake pipeline
# For historical reasons, it is only the parameters for the behavior pipeline
#
# NOTE: all hardcoded paths have been moved to the wbfm_config.yaml file
# See load_hardcoded_neural_network_paths in wbfm/utils/general/hardcoded_paths.py for more details

#
# FOR ANALYZING TRACES
#
use_barlow_tracker: true  # Experimental, possibly much better tracker

# Uncomment the relevant one; each was trained on different data
# Generally within /lisc/scratch/neurobiology/zimmer/wbfm/TrainedBarlow
# Baseline model: (2% agar)
barlow_model_path: /lisc/data/scratch/neurobiology/zimmer/MariaBoehm/barlow_track/barlow_bead_immobilized/resnet50.pth
# Fungus:
#barlow_model_path: /lisc/scratch/neurobiology/zimmer/wbfm/TrainedBarlow/hyperparameter_search_fungus/resnet50.pth
# Neuropal:
#barlow_model_path: /lisc/scratch/neurobiology/zimmer/wbfm/TrainedBarlow/hyperparameter_search_neuropal/resnet50.pth
# 1% agar:
#barlow_model_path: /lisc/scratch/neurobiology/zimmer/wbfm/TrainedBarlow/hyperparameter_search_1percent/resnet50.pth


#
# FOR ANALYZING BEHAVIOR
#
# Deletes intermediate files in the behavior pipeline
delete_intermediate_files: true

# rule subtract background
do_inverse: true

# rule normalise
alpha: 0
beta: 255

#rule stack_make_binary
threshold: 30
max_value: 255

# rule binarize_coil
coil_threshold: 240
coil_new_value: 255

# rule tiff2avi
fourcc: 0  #codec
fps: 83

# rule create centerline
nose: head    # how was the nose annotated as in DLC model? e.g. head, nose, Head, Nose, etc..
tail: tail    # how was the tail annotated as in DLC model?
num_splines: 100

# averaging spline (X,Y,K) rule
averaging_window: 83

# rule annotate behaviour  (assumes 100 segments)
initial_segment: 30
final_segment: 80
window: 100

# for turns annotation (assumes 100 segments)
turn_threshold: 0.5
turn_initial_segment: 10
turn_final_segment: 90
turn_avg_window: 500

# Hilbert
hilbert_averaging_window: 83
# Hilbert also uses the sampling frequency, which is the same as for the fft

# rule fft
sampling_frequency: 83
fft_averaging_window: 83

# Path to neural network for dlc tracking
# Comment out to use the system default (see wbfm_config.yaml)
head_tail_dlc_project: 
  /lisc/scratch/neurobiology/zimmer/zimmer_DLC/DLC_projects/DLCwbfm_headtailphar-Eva-2024-11-20/config.yaml

# rule process_centerline for equi distance
# preprocess spline
relative_spacing: 1.0  # 1% of the skeleton's total length to determine the spacing between sampled points
num_sampled_points: 10000  # Number of points to sample during spline fitting, higher values result in finer resolution
smoothing: 0.0  # Smoothing parameter for the spline fitting, controls how much the spline is smoothed (higher = more smoothing)
time_sigma: 2.0  # Sigma for Gaussian smoothing along the time (frame) axis, higher values smooth more across frames
spatial_sigma: 1.0  # Sigma for Gaussian smoothing along the spatial axis, higher values smooth more across points in each frame
max_columns: 0  # Dynamic mode: when set to 0, the skeleton is truncated where NaN content increases by more than 50%

# PCA model used for curvature analysis (should be changed for different shapes of worms)
pca_model: 
  /lisc/scratch/neurobiology/zimmer/wbfm/pca_models/2per/2per_segments_30_to_80_components_5_zscore_filtered.pkl
project_dir: 
  /lisc/data/scratch/neurobiology/zimmer/MariaBoehm/data/WBFM/projects/55_beads_barlow/2025-10-03_17-21_beads_55per_worm12-2025-10-03

**Comment by Itamar-Lev** [comment]
Important to note its a newly made barlow model, run on a new dataset

**Comment by Charles-Fieseler-Vienna** [comment]
> This is because I've been changing the default values of certain layer sizes... If you want to start from a pretrained network all the later sizes must be the same
> 
> An easy solution is to train from scratch, meaning delete the path to the pretrained model

The network you linked has a pretrained model with different settings. As far as I can see, this comment is still the solution

**Comment by Itamar-Lev** [comment]
Ok, cool cool cool, could you then expand on the explanation on where exactly it needs to be deleted?

**Comment by xmryyx** [comment]
Ok, I think I get it!
I have to set this to null in the train_config.yaml of the newly trained network and train it again, right?

/# Optional (but recommended) path to a pretrained model; note that the architecture parameters must be the same for this to work
/# If you want to train from scratch, set this to null
pretrained_model_path: /lisc/scratch/neurobiology/zimmer/wbfm/TrainedBarlow/barlow_ZIM2165_Gcamp7b_worm1-2022_11_28_from_search/trial_13/resnet50.pth  # Directly pass the .pth file of the pretrained model (the model weights)

**Comment by Charles-Fieseler-Vienna** [comment]
Yes that should be deleted. Please suggest an update to the comment there (or the documentation on the main page... but that's not really done)
